server/src/main/java/org/opensearch/search/backpressure/trackers/TaskResourceUsageTrackers.java

```diff
@@ -0,0 +1,153 @@
+        return Optional.ofNullable(all.get(type));
+    }
+
+    /**
+     * Method to access all available {@link TaskResourceUsageTracker}
+     * @return
+     */
+    public List<TaskResourceUsageTracker> all() {
+        return new ArrayList<>(all.values());
+    }
+
+    /**
+     * TaskResourceUsageTracker is used to track completions and cancellations of search related tasks.
+     * @opensearch.internal
+     */
+    public static abstract class TaskResourceUsageTracker {
+        /**
+         * Counts the number of cancellations made due to this tracker.
+         */
+        private final AtomicLong cancellations = new AtomicLong();
+        protected ResourceUsageBreachEvaluator resourceUsageBreachEvaluator;
+
+        /**
+         * for test purposes only
+         * @param resourceUsageBreachEvaluator
+         */
+        public void setResourceUsageBreachEvaluator(final ResourceUsageBreachEvaluator resourceUsageBreachEvaluator) {
+            this.resourceUsageBreachEvaluator = resourceUsageBreachEvaluator;
+        }
+
+        public long incrementCancellations() {
+            return cancellations.incrementAndGet();
+        }
+
+        public long getCancellations() {
+            return cancellations.get();
+        }
+
+        /**
+         * Returns a unique name for this tracker.
+         */
+        public abstract String name();
+
+        /**
+         * Notifies the tracker to update its state when a task execution completes.
+         */
+        public void update(Task task) {}
+
+        /**
+         * Returns the cancellation reason for the given task, if it's eligible for cancellation.
+         */
+        public Optional<TaskCancellation.Reason> checkAndMaybeGetCancellationReason(Task task) {
+            return resourceUsageBreachEvaluator.evaluate(task);
+        }
+
+        /**
+         * Returns the tracker's state for tasks as seen in the stats API.
+         */
+        public abstract Stats stats(List<? extends Task> activeTasks);
+
+        /**
+         * Method to get taskCancellations due to this tracker for the given {@link CancellableTask} tasks
+         * @param tasks
+         * @param cancellationCallback
+         * @return
+         */
+        public List<TaskCancellation> getTaskCancellations(List<CancellableTask> tasks, Runnable cancellationCallback) {
+            return tasks.stream()
+                .map(task -> this.getTaskCancellation(task, cancellationCallback))
+                .filter(TaskCancellation::isEligibleForCancellation)
+                .map(taskCancellation -> {
+                    List<Runnable> onCancelCallbacks = new ArrayList<>(taskCancellation.getOnCancelCallbacks());
+                    onCancelCallbacks.add(this::incrementCancellations);
+                    return new TaskCancellation(taskCancellation.getTask(), taskCancellation.getReasons(), onCancelCallbacks);
+                })
```

Comment:
Here we are creating this `TaskCancellation` object twice. How about doing below instead by making `getTaskCancellation` method take list of Runnable

```
return tasks.stream()
                .map(task -> this.getTaskCancellation(task, List.of(cancellationCallback, this::incrementCancellations)))
                .filter(TaskCancellation::isEligibleForCancellation)
                .collect(Collectors.toList());
```